shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala

```diff
@@ -1249,7 +1249,32 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
      */
     class ColumnarBatchToCachedBatchIterator extends InternalRowToCachedBatchIterator {
       override def getIterator: Iterator[InternalRow] = {
-        iter.asInstanceOf[Iterator[ColumnarBatch]].next.rowIterator().asScala
+
+        new Iterator[InternalRow] {
+          val batch: ColumnarBatch = iter.asInstanceOf[Iterator[ColumnarBatch]].next
+          val hostBatch = if (batch.column(0).isInstanceOf[GpuColumnVector]) {
+            withResource(batch) { batch =>
+              new ColumnarBatch(batch.safeMap(_.copyToHost()).toArray, batch.numRows())
+            }
+          } else {
+            batch
+          }
+
+          val rowIterator = hostBatch.rowIterator().asScala
+
+          override def next: InternalRow = rowIterator.next
+
+          override def hasNext: Boolean = rowIterator.hasNext
+
+          // We have to check for null context because of the unit test
+          if (TaskContext.get() != null) {
+            TaskContext.get().addTaskCompletionListener[Unit]((_: TaskContext) => {
+              if (hostBatch != null) {
```

Comment:
How is `hostBatch` going to be null?

Reply:
never. It was something I had before I decided to do it this way.